resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/K8sSubmitOps.scala

```diff
@@ -0,0 +1,192 @@
+
+import scala.collection.JavaConverters._
+
+import K8SSparkSubmitOperation.getGracePeriod
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.KubernetesClient
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.SparkSubmitOperation
+import org.apache.spark.deploy.k8s.{KubernetesUtils, SparkKubernetesClientFactory}
+import org.apache.spark.deploy.k8s.Config.{KUBERNETES_AUTH_SUBMISSION_CONF_PREFIX, KUBERNETES_SUBMIT_GRACE_PERIOD}
+import org.apache.spark.deploy.k8s.Constants.{SPARK_POD_DRIVER_ROLE, SPARK_ROLE_LABEL}
+import org.apache.spark.deploy.k8s.KubernetesUtils.formatPodState
+import org.apache.spark.util.{CommandLineLoggingUtils, Utils}
+
+private sealed trait K8sSubmitOp extends CommandLineLoggingUtils {
+  def executeOnPod(pName: String, namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit
+  def executeOnGlob(pods: List[Pod], ns: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit
+}
+
+private class KillApplication extends K8sSubmitOp  {
+  override def executeOnPod(pName: String, namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    val gracePeriodSeconds = getGracePeriod(sparkConf)
+    val podToDelete = {
+     namespace match {
+       case Some(ns) => client.pods.inNamespace(ns)
+       case None => client.pods
+     }
+    }.withName(pName)
+
+    gracePeriodSeconds match {
+      case Some(period) => podToDelete.withGracePeriod(period).delete()
+      case _ => podToDelete.delete()
+    }
+  }
+
+  override def executeOnGlob(pods: List[Pod], namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    if (pods.nonEmpty) {
+      pods.foreach { pod => printMessage(s"Deleting driver pod: ${pod.getMetadata.getName}.") }
+      val nClient = {
+        namespace match {
+          case Some(ns) => client.pods.inNamespace(ns)
+          case None => client.pods
+        }
+      }
+
+      val gracePeriodSeconds = getGracePeriod(sparkConf)
+      gracePeriodSeconds match {
+        case Some(period) =>
+          // this is not using the batch api because no option is provided
+          // when using the grace period.
+          pods.foreach { pod =>
+             nClient
+               .withName(pod.getMetadata.getName)
+               .withGracePeriod(period)
+               .delete()
+          }
+        case _ => nClient.delete(pods.asJava)
+      }
+    } else {
+      printMessage("No applications found.")
+    }
+  }
+}
+
+private class ListStatus extends K8sSubmitOp {
+  override def executeOnPod(pName: String, namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    val pod = {
+      namespace match {
+        case Some(ns) => client.pods.inNamespace(ns)
+        case None => client.pods
+      }
+    }.withName(pName)
+      .get()
```

Comment:
This looks a bit weird here. Prefer `Option(pod.get())` below.

Also, shouldn't you print a different message if the pod is not found? I see you do in `formatPodState`, but if the pod is not found I'd rather not have the header at all (like the `executeOnGlob` case).

Reply:
Fixed that.